enhancements/network/bgp-overview.md

```diff
@@ -0,0 +1,152 @@
+  - https://github.com/openshift/enhancements/pull/394
+---
+
+# OpenShift and BGP
+
+BGP (Border Gateway Protocol) is a dynamic routing protocol used to power a
+number of important networking use cases, including routing in the global
+internet.  For some helpful resources on learning more about BGP, see the
+[BGP Resources](#bgp-resources) section of this document.
+
+The purpose of this document is to serve as a high level overview of how BGP
+relates to OpenShift.  We will review a number of use cases, discuss how they
+could be supported, and link to existing related work.  Any future enhancements
+related to BGP should be able to rely on this one to provide some broader
+context and should continue to update this document as necessary.
+
+## Use Cases
+
+### L3 (Layer 3) Network Fabric
+
+The use case here does not require any changes to OpenShift.  It is still
+useful to call out as it is one use of BGP that is relevant to OpenShift,
+particularly with on premise clusters.
+
+It is possible to use BGP between the routers used to build the L3 network that
+OpenShift Nodes are connected to.
+
+### Dynamic Routing for Egress Traffic
+
+Some networks may have more than one router accessible from Nodes for
+availability reasons.  In this case an administrator may want to use BGP down
+to the Nodes themselves so they are aware of the multiple routing options and
+potentially use ECMP (Equal Cost Multi Path) to distribute traffic
+across each of the routers.
+
+This capability is not provided by OpenShift today.  It may be possible as a
+custom configuration, but not with all network providers.  With OpenShift-SDN,
+it should be possible to run your own BGP routing daemon on the host via a
+`DaemonSet`.
+
+OVN-Kuberenetes uses a different host network configuration where egress
+traffic is routed by Open vSwitch flows programmed by OVN.  This mode is
+referred to as "shared gateway mode".  It means that a dynamic routing daemon
+must also feed those routes into OVN configuration, so it's not as straight
+forward as just running an existing daemon on the host.
+
+A future enhancement is needed to describe how we would provide this
+functionality natively with OpenShift.
+
+### Pod Network Control Plane
+
+Many people become aware of BGP in the Kubernetes community via its use in
+[Calico][1].  The primary way BGP is used in Calico is as a control plane for
+the Pod network.  In other words, Calico uses BGP internally for Nodes to share
+the locations of Pod IP addresses in the cluster.
+
+You may read Calico's own documentation to understand more details about how
+Calico uses BGP.
+
+There are other ways this control plane functionality can be implemented, such
+as using the Kubernetes API or by using some other custom control plane
+technology.  Both OpenShift-SDN and OVN-Kubernetes use alternative methods to
+implement their control planes and it is not feasible to change them to use
+BGP for this purpose.
```

Comment:
I don't think that's true. The reason we avoided using BGP in OpenShift SDN in the past is that we believed that not all customers/environments would allow nodes to use BGP to configure routing, and so we'd have to continue maintaining the VXLAN-based OpenShift SDN as well, and we didn't want to maintain two different pod networking systems.

Likewise, while I don't think we have ever talked about making OVN-Kubernetes use BGP, there's no reason it _couldn't_. We'd have to figure out what to do about the loss of Geneve metadata, but there are ways we could solve that (cf. hybrid overlay).

Reply:
OK.  The mention of these other systems is probably a distraction anyway.  I think I just want to make a point that there must be *something* that implements "where is the pod IP in the cluster", and there are other ways to do it.

Reply:
take a look at the latest text and see if you think it's better